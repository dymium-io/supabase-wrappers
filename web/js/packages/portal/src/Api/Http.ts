
export function sendToServer(method: string, url: string,
    headers, body: string, onsuccess, onfailure, onexception) {
    let token = window.sessionStorage.getItem("Session");

    if (token === null) {
        onfailure(null)
        return;
    }
    let _headers = {
        Authorization: "Bearer " + token,
        Cache: "no-cache",
    }
    if (headers != null) {
        _headers = {
            ..._headers,
            ...headers
        }
    }
    let params: any = {
        method,
        headers: _headers
    }
    if (body != "") {
        params = { ...params, body: body }
    }
  

    fetch(url, params
    ).then(response => {

        if (!response.ok) {
            onfailure(response)
        } else {
            onsuccess(response)
        }
    }).catch(onexception)
}
