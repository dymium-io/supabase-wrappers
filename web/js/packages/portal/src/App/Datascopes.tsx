import React, { useEffect, useState, useRef } from 'react';
import Tabs from 'react-bootstrap/Tabs'
import Tab from 'react-bootstrap/Tab'
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Card from 'react-bootstrap/Card'
import Offcanvas from '@dymium/common/Components/Offcanvas'
import { Typeahead } from 'react-bootstrap-typeahead';
import Modal from 'react-bootstrap/Modal'
import Alert from 'react-bootstrap/Alert'
import BootstrapTable from 'react-bootstrap-table-next';
import Spinner from '@dymium/common/Components/Spinner'
import cloneDeep from 'lodash/cloneDeep';
import AddTable from './AddTable'
import EditDatascopes from './EditDatascopes'
import DatascopeForm from './DatascopeForm'
import AssignGroups from './AssignGroups'
import Groups from './Groups'
import { useAppDispatch, useAppSelector } from './hooks'
import {setActiveDatascopeTab} from '../Slices/menuSlice'

import * as com from '../Common'
import * as internal from '@dymium/common/Types/Internal'
import * as types from '@dymium/common/Types/Common'
import * as capi from '../Api/Connections'
import * as http from '@dymium/common/Api/Http'

let remap =  new internal.ConnectionMap();

export function AddDatascope(props) {
    const [validated, setValidated] = useState(false)
    let form = useRef<HTMLFormElement>(null)

    let [conns, setConns] = useState<internal.Connection[]>([])
    const [spinner, setSpinner] = useState(false)
    const [alert, setAlert] = useState<JSX.Element>(<></>)
    const [showOffcanvas, setShowOffcanvas] = useState(false)

    const [table, setTable] = useState<internal.TableScope>({schema: "", table:""})
    const [dbname, setDbname] = useState<string>("")
    const [datascope, setDatascope] = useState<internal.TablesMap>({})
    const [currentConnectionId, setCurrentConnectionId] = useState<string>("")
    const [currentConnectionType, setCurrentConnectionType] = useState<string>("")
    
    useEffect(() => {
        capi.getConnections(setSpinner, setConns, setAlert, remap, ()=>{})
    }, [])

    let sendConnection = () => {
        let retarray:types.DatascopeRecord[] = []

        Object.keys(datascope).forEach( connection => {
            let conn = datascope[connection]
            Object.keys(conn).forEach( schematable => {
                let st = conn[schematable]
          
                st.tablescope.forEach( ts => {
                    let ob:types.DatascopeRecord = types.DatascopeRecord.fromJson({connection: st.connection, schema: st.schema, table: st.table, 
                        typ: ts.typ, position: ts.position, reference: ts.reference, action:ts.action, 
                        col: ts.name, semantics: ts.semantics, dflt: ts.dflt, isnullable: ts.isnullable})
                    retarray.push(ob)
                })
            })

        } )
        // now do send
        setSpinner(true)
        let retob: types.Datascope = types.Datascope.fromJson({name: dbname, records: retarray})
        let body= retob.toJson()
        http.sendToServer("POST", "/api/savedatascope",
            null, body,
            resp => {

                resp.json().then(js => {
                    if(js.status === "OK") {
                        setAlert(
                            <Alert variant="success" onClose={() => setAlert(<></>)} dismissible>
                                Ghost Database {dbname} created successfully!
                            </Alert>
                        )
                    } else {
                        setAlert(
                            <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                                Error creating {dbname}:  {js.errormessage}!
                            </Alert>
                        )                        
                    }
                })
                setSpinner(false)
            },
            resp => {
                console.log("on error")
                setSpinner(false)
            },
            error => {
                console.log("on exception: " + error)
                setSpinner(false)
            })        
    }
    let handleSubmit = event => {
        if (form.current == null) {
            return false
        }
        if (form.current.reportValidity() === false) {
            event.preventDefault();
            setValidated(true)
            return false
        }

        event.preventDefault();
        setValidated(false)
        event.stopPropagation();

        sendConnection()

        return false
    }

    let addTableR: any = useRef(null)

    let onAddTableRef = (theref) => {
        addTableR.current = theref
    }
    let onAddTable = (table:internal.TableScope) => {
        setShowOffcanvas(false)
        if (addTableR.current !== undefined && addTableR.current.current) {
            addTableR.current.current(table)
        }
    }
    let onEditTable = (t:internal.TableScope) => {
        setTable(t)
        setShowOffcanvas(true)
    }
    let addNewTable = (id:string, dbtype:string, schema?:string, table?:string) => {
        setCurrentConnectionId(id)
        setCurrentConnectionType(dbtype)
        if(schema === undefined || table === undefined)
            setTable({ schema: "", table: "" })
        else 
            setTable({ schema, table})
        setShowOffcanvas(true)
    }
    let onDbname = (e:string) => {
        setDbname(e)
    }
    let onTablesMapUpdate = (t: internal.TablesMap) => {
        setDatascope(t)
    }
    let onDeleteConnection = (c:string) => {
        delete datascope[c]
        setDatascope(datascope)
    }
    return (
        <div className=" text-left">
            {alert}
            <Offcanvas show={showOffcanvas} onClose={(e) => { setShowOffcanvas(false) }}
                title={table["connection"] === undefined ? "Register table" : "Edit table" }>
                {showOffcanvas &&
                <AddTable onHide={() => {setShowOffcanvas(false)}} onAlert={setAlert} onAddTable={onAddTable} table={table} currentConnectionType={currentConnectionType} connectionId={currentConnectionId}/>
                }
            </Offcanvas>
            <h5 > Create New Ghost Database <Spinner show={spinner} style={{ width: '28px' }}></Spinner></h5>
            <div className=" text-left">
                <Form onSubmit={handleSubmit} ref={form} noValidate validated={validated}>
                    <DatascopeForm edit={false} dbname={dbname} onDbname={setDbname} onTablesMapUpdate={onTablesMapUpdate} 
                    onEditTable={onEditTable} AddNewTable={addNewTable} 
                    onDeleteConnection={onDeleteConnection}
                    onAddTableRef={onAddTableRef} connections={conns} setAlert={setAlert} 
                    nameToConnection={remap}/>

                    <Button data-testid="apply-datascope" variant="dymium" size="sm" className="mt-4" type="submit">
                        Apply
                    </Button>
                </Form>
            </div>
        </div>
    )
}


export default function Datascopes() {
    const t = useAppSelector((state) => {
        
        return state.reducer.activeDatascopeTab}
        )
    const appDispatch = useAppDispatch()

    return (
        <Tabs
            defaultActiveKey={t} 
            id="datascopes"
            onSelect={(k) => appDispatch( setActiveDatascopeTab(k) )}

            unmountOnExit={true} className="mb-3 text-left">
            <Tab eventKey="add" title="Add" className="mx-4">
                <AddDatascope />
            </Tab>
            <Tab eventKey="edit" title="Edit Ghost Databases" className="mx-4">
                <EditDatascopes />
            </Tab>
            <Tab eventKey="groups" title="Assign Groups" className="mx-4">
                <AssignGroups />
            </Tab>       
        </Tabs>
    )
}