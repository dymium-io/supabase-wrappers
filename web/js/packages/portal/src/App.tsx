import React, { useEffect } from 'react';
import LandingPage from './LandingPage'
import AuthenticatedApp from './App/AuthenticatedApp'
import Error404 from "./Error404"
import Logout from "./Logout"
import Error from "./Error"
import Invitation from "./Invitation"
import Sticky from "./App/Sticky"
import Dashboard from "./App/Dashboard"
import Customers from "./App/Customers"
import Connections from "./App/Connections"
import Datascopes from "./App/Datascopes"
import Groups from "./App/Groups"
import Rules from "./App/Rules"
import TestSQL from "./App/TestSQL"
import * as com from './Common'

import "@fontsource/roboto";
import 'bootstrap/dist/css/bootstrap.min.css';
import 'react-bootstrap-typeahead/css/Typeahead.css';

import { BrowserRouter, Routes, Route } from "react-router-dom";
//import Helmet from "react-helmet";

import '@dymium/common/App.scss';


function App() {
  return (
    <div className="App" id="app">

      <BrowserRouter>
        <Routes>

          <Route  path="/" element={<LandingPage />} />

          <Route path="/app/" element={<AuthenticatedApp />} >
            <Route path="/app/dashboard" element={<Dashboard />} />
            <Route path="/app/connections" element={<Connections />} />     
            <Route path="/app/datascopes" element={<Datascopes />} />                
            <Route path="/app/groups" element={<Groups />} />    
            <Route path="/app/rules" element={<Rules />} />         
            <Route path="/app/test" element={<TestSQL />} />                                             
            <Route path="/app/" element={<Sticky />} />          
          </Route>
          <Route  path="/app/logout" element={<Logout/>} />
          <Route path="/app/error" element={<Error />} />
          <Route path="/login" element={<Invitation />} />
          <Route path="*" element={<Error404 />} />
        </Routes>
      </BrowserRouter>

    </div>
  );
}

export default App;
