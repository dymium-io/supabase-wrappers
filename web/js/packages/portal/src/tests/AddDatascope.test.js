/**
 * @jest-environment jsdom
 */
import React from 'react';
import {render, fireEvent, waitFor, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from "react-router-dom";
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux'
import store from '../Store/index'
import { Http } from '../Api/Http'
import { AddDatascope } from '../App/Datascopes'
import 'bootstrap/dist/css/bootstrap.min.css';
import '@dymium/common/App.scss';
import "@fortawesome/fontawesome-free/css/all.min.css";
import mockFetch from "../mocks/mockFetch";


beforeEach(() => {
  global.fetch = jest.fn().mockImplementation(() => mockFetch)();
  global.sessionStorage.setItem("Session", "mockJWT");
  globalThis.IS_REACT_ACT_ENVIRONMENT = true;
})

afterEach(() => {
   jest.restoreAllMocks()
});

test('add-datascope', async () => {
    const component = await act( async () => render(
        <Provider store={store}>
            <BrowserRouter>
                <AddDatascope />
            </BrowserRouter>
        </Provider >
    ));

    // click on the edit button


    let connectionname =  
    await act( async () => 
        component.findByRole("textbox", {id: "dbname"} )
    )
    await act(async () => {    
        fireEvent.change(connectionname, { target: { value: "testdb" } })
    }) 

    userEvent.selectOptions(
        // Find the select element, like a real user would.
        screen.getByRole('combobox'),
        // Find and select the Ireland option, like a real user would.
        screen.getByTestId( 'adventureworks' ),
      )

/*
    //let find =  await act( async () => component.findByRole("combobox", {id: "connection"} )
    //)

    let find =  await act( async () => component.findAllByTestId('select-option')
    )    
    userEvent.selectOptions(find[0], 'adventureworks');

    let options = getAllByTestId('select-option')

    expect(options[1].selected).toBeFalsy();
    expect(options[2].selected).toBeTruthy();

*/

    let addconnection =  
    await act( async () => 
        component.findByText("Add Connection" )
    )
    await act(async () => {    
        fireEvent.click(addconnection)
    })
   
    expect(component).toMatchSnapshot()
    
    let addtable =  
    await act( async () => 
        component.findByText("Add Table" )
    )    
  
    await act(async () => {    
        fireEvent.click(addtable)
    })

    expect(component).toMatchSnapshot()



});