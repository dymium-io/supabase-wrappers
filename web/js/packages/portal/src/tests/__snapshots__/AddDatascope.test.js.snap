// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add-datascope 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" text-left"
      >
        <div
          class="bs-canvas-overlay position-fixed  w-100 h-100"
          id="bs-offcanvas-overlay"
        />
        <div
          class="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim "
          style="width: 0px; right: -600px;"
        >
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                 
                <h5>
                  Register table
                   
                </h5>
              </div>
              <div
                class="col-auto"
              >
                <i
                  aria-hidden="true"
                  class="cursor-pointer fa fa-times mr-1 fa-2x close"
                />
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="offcanvas_handle p-0 m-0 px-2 col-auto"
            />
            <div
              class="col"
            />
          </div>
        </div>
        <h5>
           Create New Data Scope 
          <img
            class="spinner"
            src="/logo.png"
            style="width: 28px;"
          />
        </h5>
        <div
          class=" text-left"
        >
          <form
            class=""
            novalidate=""
          >
            <div
              class="row"
            >
              <div
                class="col-auto"
              >
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="dbname"
                  >
                    Database Name
                  </label>
                  <input
                    class="form-control form-control-sm"
                    id="dbname"
                    pattern=".+"
                    placeholder="alphanumeric"
                    required=""
                    type="text"
                    value="testdb"
                  />
                  <div
                    class="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    class="invalid-feedback"
                  >
                    Client side name for Dymium database
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="d-flex col-auto"
                style="align-items: center;"
              >
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="connection"
                  >
                    Available Connections
                  </label>
                  <select
                    class="form-control form-control-sm"
                    id="connection"
                  >
                    <option
                      value=""
                    >
                      ...
                    </option>
                    <option
                      data-testid="northwind"
                      value="northwind"
                    >
                      northwind
                    </option>
                  </select>
                </div>
                <div
                  class="form-group"
                >
                  <label
                    class="form-label"
                  />
                  <button
                    class="btn btn-dymium btn-sm"
                    style="margin-top: 1.9em;"
                    type="button"
                  >
                    <i
                      class="fa-solid fa-database mr-2"
                    />
                    Add Connection
                  </button>
                </div>
              </div>
              <div
                class="text-left col"
              />
            </div>
            <div
              class="card mb-3 card"
              id="adventureworks"
            >
              <div
                class="card-header"
              >
                <div
                  class="row"
                >
                  <div
                    class="thickblue col-auto"
                    style="padding-top: 2px; font-size: 1.2em;"
                  >
                    <i
                      class="fa-solid fa-database mr-2"
                    />
                    Connection: 
                    adventureworks
                  </div>
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-dymium btn-sm"
                      type="button"
                    >
                      <i
                        aria-hidden="true"
                        class="fa fa-table mr-1"
                      />
                      Add Table
                    </button>
                  </div>
                  <div
                    class="text-right col-auto"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-trash blue trash"
                    />
                  </div>
                </div>
              </div>
              <div
                class="p-0 mx-0 card-body"
              />
            </div>
            <button
              class="mt-4 btn btn-dymium btn-sm"
              type="submit"
            >
              Apply
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" text-left"
    >
      <div
        class="bs-canvas-overlay position-fixed  w-100 h-100"
        id="bs-offcanvas-overlay"
      />
      <div
        class="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim "
        style="width: 0px; right: -600px;"
      >
        <div>
          <div
            class="row"
          >
            <div
              class="col"
            >
               
              <h5>
                Register table
                 
              </h5>
            </div>
            <div
              class="col-auto"
            >
              <i
                aria-hidden="true"
                class="cursor-pointer fa fa-times mr-1 fa-2x close"
              />
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="offcanvas_handle p-0 m-0 px-2 col-auto"
          />
          <div
            class="col"
          />
        </div>
      </div>
      <h5>
         Create New Data Scope 
        <img
          class="spinner"
          src="/logo.png"
          style="width: 28px;"
        />
      </h5>
      <div
        class=" text-left"
      >
        <form
          class=""
          novalidate=""
        >
          <div
            class="row"
          >
            <div
              class="col-auto"
            >
              <div
                class="mb-3 form-group"
              >
                <label
                  class="form-label"
                  for="dbname"
                >
                  Database Name
                </label>
                <input
                  class="form-control form-control-sm"
                  id="dbname"
                  pattern=".+"
                  placeholder="alphanumeric"
                  required=""
                  type="text"
                  value="testdb"
                />
                <div
                  class="valid-feedback"
                >
                  Looks good!
                </div>
                <div
                  class="invalid-feedback"
                >
                  Client side name for Dymium database
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="d-flex col-auto"
              style="align-items: center;"
            >
              <div
                class="mb-3 form-group"
              >
                <label
                  class="form-label"
                  for="connection"
                >
                  Available Connections
                </label>
                <select
                  class="form-control form-control-sm"
                  id="connection"
                >
                  <option
                    value=""
                  >
                    ...
                  </option>
                  <option
                    data-testid="northwind"
                    value="northwind"
                  >
                    northwind
                  </option>
                </select>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                />
                <button
                  class="btn btn-dymium btn-sm"
                  style="margin-top: 1.9em;"
                  type="button"
                >
                  <i
                    class="fa-solid fa-database mr-2"
                  />
                  Add Connection
                </button>
              </div>
            </div>
            <div
              class="text-left col"
            />
          </div>
          <div
            class="card mb-3 card"
            id="adventureworks"
          >
            <div
              class="card-header"
            >
              <div
                class="row"
              >
                <div
                  class="thickblue col-auto"
                  style="padding-top: 2px; font-size: 1.2em;"
                >
                  <i
                    class="fa-solid fa-database mr-2"
                  />
                  Connection: 
                  adventureworks
                </div>
                <div
                  class="col"
                >
                  <button
                    class="btn btn-dymium btn-sm"
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-table mr-1"
                    />
                    Add Table
                  </button>
                </div>
                <div
                  class="text-right col-auto"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-trash blue trash"
                  />
                </div>
              </div>
            </div>
            <div
              class="p-0 mx-0 card-body"
            />
          </div>
          <button
            class="mt-4 btn btn-dymium btn-sm"
            type="submit"
          >
            Apply
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`add-datascope 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" text-left"
      >
        <div
          class="bs-canvas-overlay-show position-fixed  w-100 h-100"
          id="bs-offcanvas-overlay"
        />
        <div
          class="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim "
          style="width: 600px; right: 0px;"
        >
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                 
                <h5>
                  Register table
                   
                </h5>
              </div>
              <div
                class="col-auto"
              >
                <i
                  aria-hidden="true"
                  class="cursor-pointer fa fa-times mr-1 fa-2x close"
                />
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="offcanvas_handle p-0 m-0 px-2 col-auto"
            />
            <div
              class="col"
            >
              <div>
                <form
                  class=""
                  novalidate=""
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-auto"
                    >
                      <div
                        class="mb-3 form-group"
                      >
                        <label
                          class="form-label"
                          for="dbname"
                        >
                          Schema Name
                        </label>
                        <div
                          class="rbt has-aux"
                          style="outline: none; position: relative;"
                          tabindex="-1"
                        >
                          <div
                            style="display: flex; flex: 1; height: 100%; position: relative;"
                          >
                            <input
                              aria-autocomplete="both"
                              aria-expanded="false"
                              aria-haspopup="listbox"
                              autocomplete="off"
                              class="rbt-input-main form-control rbt-input form-control-sm"
                              placeholder="Choose schema..."
                              role="combobox"
                              type="text"
                              value=""
                            />
                            <input
                              aria-hidden="true"
                              class="rbt-input-hint"
                              readonly=""
                              style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
                              tabindex="-1"
                              value=""
                            />
                          </div>
                          <div
                            class="rbt-aux"
                          >
                            <button
                              aria-label="Clear"
                              class="close btn-close rbt-close rbt-close-sm"
                              type="button"
                            >
                              <span
                                aria-hidden="true"
                                class="rbt-close-content"
                              >
                                ×
                              </span>
                              <span
                                class="sr-only visually-hidden"
                              >
                                Clear
                              </span>
                            </button>
                          </div>
                        </div>
                        <div
                          class="valid-feedback"
                        >
                          Looks good!
                        </div>
                        <div
                          class="invalid-feedback"
                        >
                          Client side name for Dymium database
                        </div>
                      </div>
                    </div>
                    <div
                      class="col"
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
        <h5>
           Create New Data Scope 
          <img
            class="spinner"
            src="/logo.png"
            style="width: 28px;"
          />
        </h5>
        <div
          class=" text-left"
        >
          <form
            class=""
            novalidate=""
          >
            <div
              class="row"
            >
              <div
                class="col-auto"
              >
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="dbname"
                  >
                    Database Name
                  </label>
                  <input
                    class="form-control form-control-sm"
                    id="dbname"
                    pattern=".+"
                    placeholder="alphanumeric"
                    required=""
                    type="text"
                    value="testdb"
                  />
                  <div
                    class="valid-feedback"
                  >
                    Looks good!
                  </div>
                  <div
                    class="invalid-feedback"
                  >
                    Client side name for Dymium database
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="d-flex col-auto"
                style="align-items: center;"
              >
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="connection"
                  >
                    Available Connections
                  </label>
                  <select
                    class="form-control form-control-sm"
                    id="connection"
                  >
                    <option
                      value=""
                    >
                      ...
                    </option>
                    <option
                      data-testid="northwind"
                      value="northwind"
                    >
                      northwind
                    </option>
                  </select>
                </div>
                <div
                  class="form-group"
                >
                  <label
                    class="form-label"
                  />
                  <button
                    class="btn btn-dymium btn-sm"
                    style="margin-top: 1.9em;"
                    type="button"
                  >
                    <i
                      class="fa-solid fa-database mr-2"
                    />
                    Add Connection
                  </button>
                </div>
              </div>
              <div
                class="text-left col"
              />
            </div>
            <div
              class="card mb-3 card"
              id="adventureworks"
            >
              <div
                class="card-header"
              >
                <div
                  class="row"
                >
                  <div
                    class="thickblue col-auto"
                    style="padding-top: 2px; font-size: 1.2em;"
                  >
                    <i
                      class="fa-solid fa-database mr-2"
                    />
                    Connection: 
                    adventureworks
                  </div>
                  <div
                    class="col"
                  >
                    <button
                      class="btn btn-dymium btn-sm"
                      type="button"
                    >
                      <i
                        aria-hidden="true"
                        class="fa fa-table mr-1"
                      />
                      Add Table
                    </button>
                  </div>
                  <div
                    class="text-right col-auto"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-trash blue trash"
                    />
                  </div>
                </div>
              </div>
              <div
                class="p-0 mx-0 card-body"
              />
            </div>
            <button
              class="mt-4 btn btn-dymium btn-sm"
              type="submit"
            >
              Apply
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" text-left"
    >
      <div
        class="bs-canvas-overlay-show position-fixed  w-100 h-100"
        id="bs-offcanvas-overlay"
      />
      <div
        class="bs-offcanvas position-fixed  h-100 bs-offcanvas-anim "
        style="width: 600px; right: 0px;"
      >
        <div>
          <div
            class="row"
          >
            <div
              class="col"
            >
               
              <h5>
                Register table
                 
              </h5>
            </div>
            <div
              class="col-auto"
            >
              <i
                aria-hidden="true"
                class="cursor-pointer fa fa-times mr-1 fa-2x close"
              />
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="offcanvas_handle p-0 m-0 px-2 col-auto"
          />
          <div
            class="col"
          >
            <div>
              <form
                class=""
                novalidate=""
              >
                <div
                  class="row"
                >
                  <div
                    class="col-auto"
                  >
                    <div
                      class="mb-3 form-group"
                    >
                      <label
                        class="form-label"
                        for="dbname"
                      >
                        Schema Name
                      </label>
                      <div
                        class="rbt has-aux"
                        style="outline: none; position: relative;"
                        tabindex="-1"
                      >
                        <div
                          style="display: flex; flex: 1; height: 100%; position: relative;"
                        >
                          <input
                            aria-autocomplete="both"
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            autocomplete="off"
                            class="rbt-input-main form-control rbt-input form-control-sm"
                            placeholder="Choose schema..."
                            role="combobox"
                            type="text"
                            value=""
                          />
                          <input
                            aria-hidden="true"
                            class="rbt-input-hint"
                            readonly=""
                            style="background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.54); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%; border-style: inset inset inset inset; border-width: 2px 2px 2px 2px; line-height: normal; padding: 1px 1px 1px 1px;"
                            tabindex="-1"
                            value=""
                          />
                        </div>
                        <div
                          class="rbt-aux"
                        >
                          <button
                            aria-label="Clear"
                            class="close btn-close rbt-close rbt-close-sm"
                            type="button"
                          >
                            <span
                              aria-hidden="true"
                              class="rbt-close-content"
                            >
                              ×
                            </span>
                            <span
                              class="sr-only visually-hidden"
                            >
                              Clear
                            </span>
                          </button>
                        </div>
                      </div>
                      <div
                        class="valid-feedback"
                      >
                        Looks good!
                      </div>
                      <div
                        class="invalid-feedback"
                      >
                        Client side name for Dymium database
                      </div>
                    </div>
                  </div>
                  <div
                    class="col"
                  />
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
      <h5>
         Create New Data Scope 
        <img
          class="spinner"
          src="/logo.png"
          style="width: 28px;"
        />
      </h5>
      <div
        class=" text-left"
      >
        <form
          class=""
          novalidate=""
        >
          <div
            class="row"
          >
            <div
              class="col-auto"
            >
              <div
                class="mb-3 form-group"
              >
                <label
                  class="form-label"
                  for="dbname"
                >
                  Database Name
                </label>
                <input
                  class="form-control form-control-sm"
                  id="dbname"
                  pattern=".+"
                  placeholder="alphanumeric"
                  required=""
                  type="text"
                  value="testdb"
                />
                <div
                  class="valid-feedback"
                >
                  Looks good!
                </div>
                <div
                  class="invalid-feedback"
                >
                  Client side name for Dymium database
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="d-flex col-auto"
              style="align-items: center;"
            >
              <div
                class="mb-3 form-group"
              >
                <label
                  class="form-label"
                  for="connection"
                >
                  Available Connections
                </label>
                <select
                  class="form-control form-control-sm"
                  id="connection"
                >
                  <option
                    value=""
                  >
                    ...
                  </option>
                  <option
                    data-testid="northwind"
                    value="northwind"
                  >
                    northwind
                  </option>
                </select>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                />
                <button
                  class="btn btn-dymium btn-sm"
                  style="margin-top: 1.9em;"
                  type="button"
                >
                  <i
                    class="fa-solid fa-database mr-2"
                  />
                  Add Connection
                </button>
              </div>
            </div>
            <div
              class="text-left col"
            />
          </div>
          <div
            class="card mb-3 card"
            id="adventureworks"
          >
            <div
              class="card-header"
            >
              <div
                class="row"
              >
                <div
                  class="thickblue col-auto"
                  style="padding-top: 2px; font-size: 1.2em;"
                >
                  <i
                    class="fa-solid fa-database mr-2"
                  />
                  Connection: 
                  adventureworks
                </div>
                <div
                  class="col"
                >
                  <button
                    class="btn btn-dymium btn-sm"
                    type="button"
                  >
                    <i
                      aria-hidden="true"
                      class="fa fa-table mr-1"
                    />
                    Add Table
                  </button>
                </div>
                <div
                  class="text-right col-auto"
                >
                  <i
                    aria-hidden="true"
                    class="fa fa-trash blue trash"
                  />
                </div>
              </div>
            </div>
            <div
              class="p-0 mx-0 card-body"
            />
          </div>
          <button
            class="mt-4 btn btn-dymium btn-sm"
            type="submit"
          >
            Apply
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
